Realm/Tests/NotificationTests.m

```diff
@@ -1566,4 +1585,145 @@ - (void)testDiffedUpdatesOnlyNotifyForPropertiesWhichActuallyChanged {
+    [self waitForExpectationsWithTimeout:0.1 handler:nil];
+
+    [token invalidate];
+}
+
+- (void)testModifyObservedKeyPathRemotely {
+    XCTestExpectation *ex = [self expectationWithDescription:@"change notification"];
+    RLMNotificationToken *token = [_obj addNotificationBlock:^(BOOL deleted, NSArray *changes, NSError *error) {
+        XCTAssertFalse(deleted);
+        XCTAssertNil(error);
+        XCTAssertEqual(changes.count, 1U);
+        RLMPropertyChange *prop = changes[0];
+        XCTAssertEqualObjects(prop.name, @"boolCol");
+        [ex fulfill];
+    } keyPaths:@[@"boolCol"]];
+
+    [self dispatchAsync:^{
+        RLMRealm *realm = [RLMRealm defaultRealm];
+        AllTypesObject *obj = [[AllTypesObject allObjectsInRealm:realm] firstObject];
+        [realm beginWriteTransaction];
+        XCTAssertNotEqual(obj.boolCol, [_values[@"boolCol"] boolValue]);
+        obj.boolCol = [_values[@"boolCol"] boolValue];
+        [realm commitWriteTransaction];
+    }];
+    [self waitForExpectationsWithTimeout:0.1 handler:nil];
+
+    [token invalidate];
+}
+
+- (void)testModifyUnobservedKeyPathRemotely {
+    XCTestExpectation *ex = [self expectationWithDescription:@"no change notification"];
+    ex.inverted = true;
+
+    RLMNotificationToken *token = [_obj addNotificationBlock:^(__unused BOOL deletd,
+                                                               __unused NSArray<RLMPropertyChange *> *changes,
+                                                               __unused NSError *error) {
+        [ex fulfill];
+    } keyPaths:@[@"boolCol"]];
+
+    [self dispatchAsync:^{
+        RLMRealm *realm = [RLMRealm defaultRealm];
+        AllTypesObject *obj = [[AllTypesObject allObjectsInRealm:realm] firstObject];
+        [realm beginWriteTransaction];
+        obj.intCol = obj.intCol + obj.intCol;
+        [realm commitWriteTransaction];
+    }];
+    [self waitForExpectationsWithTimeout:0.1 handler:nil];
+
+    [token invalidate];
+}
+
+- (void)testModifyObservedKeyPathArrayProperty {
+    XCTestExpectation *ex = [self expectationWithDescription:@"change notification"];
+
+    RLMRealm *realm = RLMRealm.defaultRealm;
+    [realm beginWriteTransaction];
+    CompanyObject *company = [CompanyObject createInRealm:realm withValue:@{}];
+    EmployeeObject *employee = [EmployeeObject createInRealm:realm withValue:@{@"age": @30, @"hired": @NO}];
+    [company.employees addObject:employee];
+    [realm commitWriteTransaction];
+
+    RLMNotificationToken *token = [company addNotificationBlock:^(BOOL deleted, NSArray *changes, NSError *error) {
+        XCTAssertFalse(deleted);
+        XCTAssertNil(error);
+        XCTAssertEqual(changes.count, 1U);
+
+        for (RLMPropertyChange *prop in changes) {
+            XCTAssertEqualObjects(prop.name, @"employees");
+            XCTAssertFalse(prop.previousValue);
+            XCTAssertFalse(prop.value); // Observing an array will lead to NSNull here.
```

Comment:
Either the test or the comment is wrong here? If `prop.value` actually is NSNull rather than nil then it'll evaluate to true when cast to a bool.

Reply:
The comment is wrong.